### PR TITLE
Python Interface

### PR DESCRIPTION
**Pull request by bcumming**
A big update that wraps over 90% of the public C++ API, with enough functionality to let Python users to perform useful modelling.

Key features
- wrapping of cable cell functionality
  - user-defined explicit compartmentalisation not supported
- `single_cell_model` abstraction unique to the Python wrapper that simplifies developing and testing single cell models
- one-dimensional cell builder for Python wrapper that simplfies building cells that 
  - in and of itself limited in scope, but a very useful example of mapping a richer morphology builder onto `sample_tree`s.
- parsing of region and location expressions from strings
  - implements a generic s-expression parser that we can use for other purposes later

### CODE REVIEW

File: python/morphology.cpp [hunk 1]
```diff
@@ -24,10 +36,178 @@ void register_morphology(pybind11::module& m) {
...
+            [](const arb::msample& s) {
+                return util::pprintf("<arbor.mpoint: x {}, y {}, z {}, radius {}, tag {}>",
+                        s.loc.x, s.loc.y, s.loc.z, s.loc.radius, s.tag); })
+        .def("__repr__",
+            [](const arb::msample& s) {
+                return util::pprintf("{}>", s);});
+
+    // arb::mcable
+    pybind11::class_<arb::mcable> cable(m, "cable");
+    cable
```
Comment (akuesters) [inline python/morphology.cpp]: add description, e.g. `An unbranched cable in the morphology. `

File: python/morphology.cpp [hunk 2]
```diff
@@ -10,7 +16,13 @@ namespace pyarb {
...
-    // segment location
+    //
+    //  primitives: points, samples, locations, cables... etc.
+    //
+
+    m.attr("mnpos") = arb::mnpos;
+
+    // arb::mlocation
     pybind11::class_<arb::mlocation> location(m, "location",
         "A location on a cable cell.");
```
Comment (akuesters) [inline python/morphology.cpp]: rather `A specific location on a morpholology.`

File: python/morphology.cpp [hunk 3]
```diff
@@ -24,10 +36,178 @@ void register_morphology(pybind11::module& m) {
...
+    cable
+        .def(pybind11::init(
+                    [](arb::msize_t bid, double prox, double dist) {
+                        arb::mcable c{bid, prox, dist};
+                        if (!test_invariants(c)) {
+                            throw pyarb_error("Invalid cable description. Cable segments must have proximal and distal end points in the range [0,1].");
+                        }
+                        return c;
+                    }),
+             "branch_id"_a, "prox"_a, "dist"_a)
```
Comment (akuesters) [inline python/morphology.cpp]: add description, e.g. `Create an unbranched cable in the morphology with the id of the branch on which the cable lies, the relative position of the proximal end of the cable on its branch and the relative position of the distal end of the cable on its branch.`

File: python/morphology.cpp [hunk 4]
```diff
@@ -24,10 +36,178 @@ void register_morphology(pybind11::module& m) {
...
+        .def_readonly("prox_pos", &arb::mcable::prox_pos,
+                "The relative position of the proximal end of the cable on its branch ∈ [0,1].")
+        .def_readonly("dist_pos", &arb::mcable::dist_pos,
+                "The relative position of the distal end of the cable on its branch ∈ [0,1].")
+        .def_readonly("branch", &arb::mcable::branch,
+                "The id of the branch on which the cable lies.")
+        .def("__str__", [](const arb::mcable& c) {
```
Comment (akuesters) [inline python/morphology.cpp]: make consistent (i.e., repr with <arbor.class: ...>, str without)

File: python/morphology.cpp [hunk 5]
```diff
@@ -24,10 +36,178 @@ void register_morphology(pybind11::module& m) {
...
+            return util::pprintf("<arbor.cable: branch {}, prox {}, dist {}", c.branch, c.prox_pos, c.dist_pos); })
+        .def("__repr__", [](const arb::mcable& c) { return util::pprintf("{}", c); });
+
+    //
+    // Higher-level data structures (sample_tree, morphology)
+    //
+
+    // arb::sample_tree
+    pybind11::class_<arb::sample_tree> sample_tree(m, "sample_tree");
+    sample_tree
```
Comment (akuesters) [inline python/morphology.cpp]: add description on what a sample_tree is, e.g. `A morphology composed of samples.`

File: python/morphology.cpp [hunk 6]
```diff
@@ -24,10 +36,178 @@ void register_morphology(pybind11::module& m) {
...
+        // constructors
+        .def(pybind11::init<>())
+        // modifiers
+        .def("reserve", &arb::sample_tree::reserve)
```
Comment (akuesters) [inline python/morphology.cpp]: add description, e.g. `Reserve space for n samples.`,

File: python/morphology.cpp [hunk 7]
```diff
@@ -24,10 +36,178 @@ void register_morphology(pybind11::module& m) {
...
+        .def("append", [](arb::sample_tree& t, arb::msample s){t.append(s);})
```
Comment (akuesters) [inline python/morphology.cpp]: add description, e.g. `Append a single sample to the last sample in the tree.`

File: python/morphology.cpp [hunk 8]
```diff
@@ -24,10 +36,178 @@ void register_morphology(pybind11::module& m) {
...
+        .def("append", [](arb::sample_tree& t, arb::msize_t p, arb::msample s){t.append(p, s);})
```
Comment (akuesters) [inline python/morphology.cpp]: add description, e.g. `Append a single sample to another sample.`

File: python/morphology.cpp [hunk 9]
```diff
@@ -24,10 +36,178 @@ void register_morphology(pybind11::module& m) {
...
+        // properties
+        .def_property_readonly("empty", [](const arb::sample_tree& st){return st.empty();},
+                "Indicates whether the sample tree is empty (i.e. whether it has size 0)")
```
Comment (akuesters) [inline python/morphology.cpp]: add a dot at the end of the sentence.

File: python/morphology.cpp [hunk 10]
```diff
@@ -24,10 +36,178 @@ void register_morphology(pybind11::module& m) {
...
+        .def_property_readonly("size", [](const arb::sample_tree& st){return st.size();},
+                "The number of samples in the sample tree.")
+        .def_property_readonly("parents", [](const arb::sample_tree& st){return st.parents();},
+                "A list with the parent index of each sample.")
+        .def("__str__", [](const arb::sample_tree& s) {
```
Comment (akuesters) [inline python/morphology.cpp]: add `__repr__` and make consistent

File: python/morphology.cpp [hunk 11]
```diff
@@ -24,10 +36,178 @@ void register_morphology(pybind11::module& m) {
...
+
+    // arb::morphology
+
+    pybind11::class_<arb::morphology> morph(m, "morphology");
+    morph
+        // constructors
+        .def(pybind11::init(
+                [](arb::sample_tree t){
+                    return arb::morphology(std::move(t));
+                }))
```
Comment (akuesters) [inline python/morphology.cpp]: Add description, e.g. `Create a morphology from a sample tree.`

File: python/morphology.cpp [hunk 12]
```diff
@@ -24,10 +36,178 @@ void register_morphology(pybind11::module& m) {
...
+        .def(pybind11::init(
+                [](arb::sample_tree t, bool spherical_root) {
+                    return arb::morphology(std::move(t), spherical_root);
+                }), "sample_tree"_a, "spherical_root"_a)
```
Comment (akuesters) [inline python/morphology.cpp]: Add description, e.g. `Create a morphology from a sample tree with a spherical root (true/false).`

File: python/morphology.cpp [hunk 13]
```diff
@@ -24,10 +36,178 @@ void register_morphology(pybind11::module& m) {
...
+        // morphology's interface is read-only by design, so most of it can
+        // be implemented as read-only properties.
+        .def_property_readonly("empty",
+                [](const arb::morphology& m){return m.empty();},
+                "A list with the parent index of each sample.")
```
Comment (akuesters) [inline python/morphology.cpp]: description doesn't fit.

File: python/morphology.cpp [hunk 14]
```diff
@@ -24,10 +36,178 @@ void register_morphology(pybind11::module& m) {
...
+        .def("branch_indexes",
+                [](const arb::morphology& m, arb::msize_t i) {
+                    auto p = m.branch_indexes(i);
+                    return std::vector<arb::msize_t>(p.first, p.second);
+                },
+                "i"_a, "Range of indexes into the sample points in branch i.")
+        .def("__str__",
+                [](const arb::morphology& m) {
+                    return util::pprintf("<arbor.morphology:\n{}>", m);
+                });
```
Comment (akuesters) [inline python/morphology.cpp]: add `__repr__` nd make consistent

File: python/s_expr.cpp [hunk 15]
```diff
@@ -0,0 +1,404 @@
+#include <iostream>
+
+#include <cctype>
+#include <cstring>
+#include <string>
+#include <memory>
+#include <ostream>
+#include <vector>
```
Comment (akuesters) [inline python/s_expr.cpp]: not alphabetical

File: python/s_expr.cpp [hunk 16]
```diff
@@ -0,0 +1,404 @@
...
+    return state.get<1>().tail.get();
+}
+
+s_expr::operator bool() const {
+    return !(is_atom() && atom().kind==tok::nil);
+}
+
+std::ostream& operator<<(std::ostream& o, const s_expr& x) {
+    if (x.is_atom()) return o << x.atom();
+#if 0 // print full tree with terminating 'nil'
```
Comment (akuesters) [inline python/s_expr.cpp]: are those supposed to be commented out (with #)?

File: python/single_cell_model.cpp [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+    const std::vector<trace>& traces() const {
+        return traces_;
+    }
+};
+
+void register_single_cell(pybind11::module& m) {
+    using namespace pybind11::literals;
+
+    pybind11::class_<trace> tr(m, "trace");
+    tr
```
Comment (akuesters) [inline python/single_cell_model.cpp]: add description on what a trace is

File: python/single_cell_model.cpp [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+        .def_readonly("variable", &trace::variable)
```
Comment (akuesters) [inline python/single_cell_model.cpp]: add descriptions to properties (use capital letters and no dot at the end of the sentence.)

File: python/single_cell_model.cpp [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+        .def_readonly("location", &trace::loc)
+        .def_readonly("time",    &trace::t)
+        .def_readonly("value",   &trace::v);
+
+    pybind11::class_<single_cell_model> model(m, "single_cell_model",
+        "Wrapper for easily building and running single cell models.");
+
```
Comment (akuesters) [inline python/single_cell_model.cpp]: rm line

File: python/single_cell_model.cpp [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+    model
+        .def(pybind11::init<arb::cable_cell>(),
+            "cell"_a, "Build a single cell model.")
+        .def("run", &single_cell_model::run, "tfinal"_a, "run model from t=0 to t=tfinal ms")
```
Comment (akuesters) [inline python/single_cell_model.cpp]: `Run model from t=0 to t=tfinal ms.`

File: python/single_cell_model.cpp [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+        .def("probe", &single_cell_model::probe, "what"_a, "where"_a, "frequency"_a)
```
Comment (akuesters) [inline python/single_cell_model.cpp]: add description, e.g. `Attach probes sampling "what" at location "where" at frequency.`

File: python/single_cell_model.cpp [hunk 22]
```diff
@@ -0,0 +1,230 @@
...
+        .def_property_readonly("spikes", [](const single_cell_model& m) {return m.spike_times();}, "spike times (ms)")
```
Comment (akuesters) [inline python/single_cell_model.cpp]: capital letters, build sentences, use [] for units e.g.`Return spike times [ms].`

File: python/cells.cpp [hunk 23]
```diff
@@ -216,47 +221,281 @@ void register_cells(pybind11::module& m) {
...
+    pybind11::class_<arb::cable_cell> cable_cell(m, "cable_cell");
     cable_cell
+        .def(pybind11::init(
+            [](const arb::morphology& m, const label_dict_proxy& labels) {
+                return arb::cable_cell(m, labels.dict);
+            }), "morphology"_a, "labels"_a)
+        .def(pybind11::init(
+            [](const arb::sample_tree& t, const label_dict_proxy& labels) {
+                return arb::cable_cell(arb::morphology(t), labels.dict);
+            }), "morphology"_a, "labels"_a)
```
[the remark this answers is not in the document]
Comment (bcumming) [reply python/cells.cpp]: I keep it as `morphology`, and add a docstring stating that the `sample_tree` is used to build the morphology.

File: python/example/single_cell_builder.py [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+cell.place('stim_site', arbor.iclamp( 10, 2, 0.8))
+cell.place('stim_site', arbor.iclamp( 50, 2, 0.8))
+cell.place('stim_site', arbor.iclamp( 80, 2, 0.8))
+cell.place('root', arbor.spike_detector(-10))
+
+# make single cell model
+m = arbor.single_cell_model(cell)
+
+# Attach voltage probes, sampling at 10 kHz.
+m.probe('voltage', loc(0,0), 10000) # at the soma
```
[the remark this answers is not in the document]
Comment (bcumming) [reply python/example/single_cell_builder.py]: I think that the comments above make it pretty clear what is going on.

File: python/example/single_cell_swc.py [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+cell.place('root', arbor.spike_detector(-10))
+
+# Have one compartment between each sample point.
+cell.compartments_on_samples()
+
+# Make single cell model.
+m = arbor.single_cell_model(cell)
+
+# Attach voltage probes that sample at 50 kHz.
+m.probe('voltage', where=loc(0,0),  frequency=50000)
```
[the remark this answers is not in the document]
Comment (akuesters) [reply python/example/single_cell_swc.py]: still missing location as in `single_cell_builder.py`

File: python/flat_cell_builder.cpp [hunk 26]
```diff
@@ -0,0 +1,272 @@
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include <arbor/morph/morphology.hpp>
+#include <arbor/morph/primitives.hpp>
+#include <arbor/morph/sample_tree.hpp>
+#include <arbor/cable_cell.hpp>
```
[the remark this answers is not in the document]
Comment (akuesters) [reply python/flat_cell_builder.cpp]: c before m :)

File: python/flat_cell_builder.cpp [hunk 27]
```diff
@@ -0,0 +1,272 @@
...
+        if (!at_root && parent>=size()) {
+            throw pyarb_error("Invalid parent id.");
+        }
+
+        // Calculating the sample that is the parent of this branch is
+        // neccesarily complicated to handle cable segments that are attached
+        // to the root.
+        arb::msize_t p = at_root? (size()? 0: mnpos): cable_distal_id_[parent];
+
+        //std::cout << " Parent " << p << "\n";
```
[the remark this answers is not in the document]
Comment (akuesters) [reply python/flat_cell_builder.cpp]: Don't we want to remove these?

File: python/flat_cell_builder.cpp [hunk 28]
```diff
@@ -0,0 +1,272 @@
...
+                   tree_.samples()[p].loc.z;        // attach to end of a cable
+
+        // Only add a first point at the very beginning of the cable if
+        // the cable is not attached to another
+        const bool add_first_point = p==arb::mnpos      // attached to the "root"
+                                  || (!p && spherical_) // attached to a spherical root
+                                  || (r1!=tree_.samples()[p].loc.radius);
+                                                        // proximal radius does not match r1
+        if (add_first_point) {
+            //std::cout << "   Up front\n";
```
[the remark this answers is not in the document]
Comment (akuesters) [reply python/flat_cell_builder.cpp]: see above

File: python/flat_cell_builder.cpp [hunk 29]
```diff
@@ -0,0 +1,272 @@
...
+            return tag_count_;
+        }
+    }
+
+    // Only valid if called on a non-empty tree with spherical soma.
+    double soma_rad() const {
+        return tree_.samples()[0].loc.radius;
+    }
+
+    // The number of cable segements (plus one optional soma) were used to
```
[the remark this answers is not in the document]
Comment (akuesters) [reply python/flat_cell_builder.cpp]: typo

File: python/flat_cell_builder.cpp [hunk 30]
```diff
@@ -0,0 +1,272 @@
...
+    std::size_t size() const {
+        return cable_distal_id_.size();
+    }
+};
+
+void register_flat_builder(pybind11::module& m) {
+    using namespace pybind11::literals;
+
+    pybind11::class_<flat_cell_builder> builder(m, "flat_cell_builder");
+    builder
```
[the remark this answers is not in the document]
Comment (akuesters) [reply python/flat_cell_builder.cpp]: review seems not to be applied past this point?

File: doc/python.rst [hunk 31]
```diff
@@ -0,0 +1,48 @@
+.. _getstarted_python:
+
+Python
+======
+
+Arbor provies access to all of the C++ library's functionality in Python,
+which is the only interface that the vast majority of users will use.
+The getting started guides will introduce Arbor using the Python.
```
[the remark this answers is not in the document]
Comment (bcumming) [reply doc/python.rst]: These docs were accidentally merged into this PR, and I left them here as placeholders.
The other guides don't exist yet (they will be written next).

File: doc/single_cell.rst [hunk 32]
```diff
@@ -0,0 +1,90 @@
...
+    in NEURON, which conincdently has the same surface area as a sphere of the same diameter.
+    Arbor allows the user to optionally use a spherical section at the root
+    of the tree to represent spherical somata.
+
+.. note::
+    In NEURON cell morphologies are constructed by creating individual sections,
+    then connecting them together. In Arbor we start with an "empty"
+    sample tree, to which samples are appended to build a connected morphology.
+
+Let's start with a simple "ball and stick" model cell.
```
[the remark this answers is not in the document]
Comment (bcumming) [reply doc/single_cell.rst]: Maybe a bit informal, but should be ok? I will reword this on the next pass anyway.

File: python/cells.cpp [hunk 33]
```diff
@@ -216,47 +221,326 @@ void register_cells(pybind11::module& m) {
...
+
+    // arb::i_clamp
+
+    pybind11::class_<arb::i_clamp> i_clamp(m, "iclamp",
+        "A current clamp, for injecting a single pulse of current with fixed duration and current.");
+    i_clamp
+        .def(pybind11::init(
+                [](double ts, double dur, double cur) {
+                    return arb::i_clamp{ts, dur, cur};
+                }), "tstart"_a=0, "duration"_a=0, "current"_a=0)
```
[the remark this answers is not in the document]
Comment (bcumming) [reply python/cells.cpp]: I don't thing that they need them for such obvious cases, where we have named parameters and the rest of the interface defines their meaning.
I would reserve such descriptions for when they are required.

File: python/cells.cpp [hunk 34]
```diff
@@ -216,47 +221,326 @@ void register_cells(pybind11::module& m) {
...
+            "ion"_a,
+            "int_con"_a=pybind11::none(),
+            "ext_con"_a=pybind11::none(),
+            "rev_pot"_a=pybind11::none(),
+            "method"_a=pybind11::none(),
+            "Set the propoerties of ion species named 'ion' that will be applied\n"
+            "by default everywhere on the cell. Species concentrations and reversal\n"
+            "potential can be overridden on specific regions using the paint interface, \n"
+            "while the method for calculating reversal potential is global for all\n"
+            "compartments in the cell, and can't be overriden locally.\n"
```
[the remark this answers is not in the document]
Comment (bcumming) [reply python/cells.cpp]: This is fine, in my opinion.

File: python/cells.cpp [hunk 35]
```diff
@@ -216,47 +221,326 @@ void register_cells(pybind11::module& m) {
...
+        "A current clamp, for injecting a single pulse of current with fixed duration and current.");
+    i_clamp
+        .def(pybind11::init(
+                [](double ts, double dur, double cur) {
+                    return arb::i_clamp{ts, dur, cur};
+                }), "tstart"_a=0, "duration"_a=0, "current"_a=0)
+        .def_readonly("tstart", &arb::i_clamp::delay,       "Time at which current starts [ms]")
+        .def_readonly("duration", &arb::i_clamp::duration,  "Duration of the current injection [ms]")
+        .def_readonly("current", &arb::i_clamp::amplitude,  "Amplitude of the injected current [nA]")
+        .def("__repr__", [](const arb::i_clamp& c){
```
[the remark this answers is not in the document]
Comment (bcumming) [reply python/cells.cpp]: There are many existing cases where this is not consistent. Let's save this for when we go through the whole interface.

**Pull request merged**